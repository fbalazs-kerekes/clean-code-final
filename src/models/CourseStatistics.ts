export class CourseStatistics {
    constructor(
        private courseName: string,
        private totalLectures: number,
        private lecturesCompleted: number,
        private progress: number,
        private lastAccessed: Date
    ) { }

    GetCourseName() {
        return this.courseName;
    }
}