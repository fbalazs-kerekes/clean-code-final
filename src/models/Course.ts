import { Student } from "./Student";

export class Course {
    private students: Student[] = [];
    constructor(private courseName: string, private startDate: Date, private lengthInWeeks: number, private costInHuf: number) { }

    public GetCourseName(): string {
        return this.courseName;
    }

    public GetStartDate(): Date {
        return this.startDate;
    }

    public GetLengthInWeeks(): number {
        return this.lengthInWeeks;
    }

    public GetCostInHuf(): number {
        return this.costInHuf;
    }

    public GetStudents(): Student[] {
        return this.students;
    }
}
