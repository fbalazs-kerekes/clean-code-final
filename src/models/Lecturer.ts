import { Course } from "./Course";
import { Person } from "./Person";

export class Lecturer extends Person {
    private assignedCourses: Course[] = [];
}