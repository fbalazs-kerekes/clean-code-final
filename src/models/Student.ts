import { Course } from "./Course";
import { Person } from "./Person";

export class Student extends Person {
    private registredCourses: Course[] = [];
}