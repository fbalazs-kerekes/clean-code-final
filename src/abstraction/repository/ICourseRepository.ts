import { Course } from "../../models/Course";
import { CourseStatistics } from "../../models/CourseStatistics";
import { Student } from "../../models/Student";

export interface ICourseRepository {
    AddCourse(course: Course): Promise<void>;
    AddStudentToCourse(student: Student, course: Course): Promise<void>;
    GetCourseByName(courseName: string): Promise<Course | undefined>;
    GetCourses(): Promise<Course[]>;
    GetCourseStatistics(course: Course): Promise<CourseStatistics | undefined>;
}