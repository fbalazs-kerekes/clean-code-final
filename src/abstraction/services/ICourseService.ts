import { Course } from "../../models/Course";
import { CourseStatistics } from "../../models/CourseStatistics";
import { Student } from "../../models/Student";

export interface ICourseService {
    AddCourse(course: Course): Promise<void>;
    GetCourses(): Promise<Course[]>;
    GetCourseByName(name: string): Promise<Course | undefined>;
    AddStudentToCourse(student: Student, course: Course): Promise<void>;
    GetCourseStatistics(course: Course): Promise<CourseStatistics | undefined>;
}