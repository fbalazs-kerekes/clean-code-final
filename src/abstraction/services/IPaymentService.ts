import { Student } from "../../models/Student";

export interface IPaymentService {
    isOrderPaid(student: Student): Promise<boolean>;
    
}