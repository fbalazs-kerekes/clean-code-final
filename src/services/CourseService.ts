import { Course } from "../models/Course";
import { Student } from "../models/Student";
import { ICourseRepository } from "../abstraction/repository/ICourseRepository";
import { IPaymentService } from "../abstraction/services/IPaymentService";
import { INotificationService } from "../abstraction/services/INotificationService";
import { ICourseService } from "../abstraction/services/ICourseService";
import { ResourceNotFoundException } from "../exceptions/ResourceNotFoundException";
import { PaymentException } from "../exceptions/PaymentException";
import { CourseStatistics } from "../models/CourseStatistics";

export class CourseService implements ICourseService {
    constructor(
        private courseRepository: ICourseRepository,
        private paymentService: IPaymentService,
        private notificationService: INotificationService) { }

    public async AddCourse(course: Course): Promise<void>{
        await this.courseRepository.AddCourse(course);
    }

    public async GetCourses(): Promise<Course[]> {
        return await this.courseRepository.GetCourses();
    }

    public async GetCourseByName(courseName: string): Promise<Course | undefined> {

        const course = await this.courseRepository.GetCourseByName(courseName);
        if (!course) { throw new ResourceNotFoundException('Course not found') }

        return course;
    }

    public async AddStudentToCourse(student: Student, course: Course): Promise<void>{
        await this.validateOrderPayed(student);

        await this.courseRepository.AddStudentToCourse(student, course);
        await this.notificationService.SendNotifications(`${student.GetName()} student was added to course.`)
    }

    public async validateOrderPayed(student: Student): Promise<void> {
        const isCoursePayedByStudent = await this.paymentService.isOrderPaid(student);
        if (!isCoursePayedByStudent) { throw new PaymentException('Course is not yet paid by Student.') };
    }

    public async GetCourseStatistics(course: Course): Promise<CourseStatistics | undefined> {
        return await this.courseRepository.GetCourseStatistics(course);
    }
}