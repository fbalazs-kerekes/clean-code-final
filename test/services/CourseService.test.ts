import { CourseService } from "../../src/services/CourseService";
import { mock, mockReset } from "jest-mock-extended";
import { Course } from "../../src/models/Course";
import { DatabaseException } from "../../src/exceptions/DatabaseException";
import { Student } from "../../src/models/Student";
import { PaymentException } from "../../src/exceptions/PaymentException";
import { CourseStatistics } from "../../src/models/CourseStatistics";
import { ResourceNotFoundException } from "../../src/exceptions/ResourceNotFoundException";
import { UnknownErrorException } from "../../src/exceptions/UnknownErrorException";
import { ICourseService } from "../../src/abstraction/services/ICourseService";
import { ICourseRepository } from "../../src/abstraction/repository/ICourseRepository";
import { IPaymentService } from "../../src/abstraction/services/IPaymentService";
import { INotificationService } from "../../src/abstraction/services/INotificationService";

const mockCourseRepository = mock<ICourseRepository>();
const mockPaymentService = mock<IPaymentService>();
const mockNotificationService = mock<INotificationService>();

const courses: Course[] = [
    new Course('CR1', new Date('2024-01-01'), 5, 10000),
    new Course('CR2', new Date('2024-02-01'), 4, 20000),
    new Course('CR3', new Date('2024-03-01'), 3, 30000),
];

const courseStatistics = [
    new CourseStatistics('CR1', 100, 50, 30, new Date('2024-01-02 11:10:00')),
    new CourseStatistics('CR2', 200, 50, 30, new Date('2024-02-02 11:10:00')),
    new CourseStatistics('CR3', 150, 50, 30, new Date('2024-03-02 11:10:00'))
];

describe('CourseService test', () => {
    let courseService: ICourseService;

    beforeEach(() => {
        mockReset(mockCourseRepository);
        mockReset(mockPaymentService);
        mockReset(mockNotificationService);
        courseService = new CourseService(mockCourseRepository, mockPaymentService, mockNotificationService)
    })

    describe('Happy path', () => {

        const course = new Course('C4', new Date('2024-04-01'), 5, 40000);
        const student = new Student('Student 1', 20);

        it('should add course into database', async () => {
            mockCourseRepository.AddCourse.mockResolvedValue(void 0);

            await courseService.AddCourse(course);

            expect(mockCourseRepository.AddCourse).toHaveBeenCalledTimes(1);
            expect(mockCourseRepository.AddCourse).toHaveBeenCalledWith(course);
        })

        it('should add student to course into database', async () => {
            mockCourseRepository.AddCourse.mockResolvedValue(void 0);
            mockPaymentService.isOrderPaid.mockResolvedValue(true);

            await courseService.AddStudentToCourse(student, course);

            expect(mockCourseRepository.AddStudentToCourse).toHaveBeenCalledTimes(1);
            expect(mockCourseRepository.AddStudentToCourse).toHaveBeenCalledWith(student, course);
        })

        it('should get all course from database', async () => {
            mockCourseRepository.GetCourses.mockResolvedValue(courses);

            let courseServiceCourses = await courseService.GetCourses();

            expect(courseServiceCourses).toMatchSnapshot();
            expect(mockCourseRepository.GetCourses).toHaveBeenCalledTimes(1);
        })

        it.each(courses)('should get course by name ($courseName) from database', async (expectedCourse: Course) => {
            const courseName = expectedCourse.GetCourseName();
            mockCourseRepository.GetCourseByName.mockResolvedValue(expectedCourse);

            let courseElement = await courseService.GetCourseByName(courseName);

            expect(courseElement).toBe(expectedCourse);
            expect(mockCourseRepository.GetCourseByName).toHaveBeenCalledTimes(1);
            expect(mockCourseRepository.GetCourseByName).toHaveBeenCalledWith(courseName);
        })

        it.each(courses)('should get resource not found exception when course does not exists by name ($courseName) from database', async (expectedCourse: Course) => {
            const courseName = expectedCourse.GetCourseName();
            const expectedErrorMessage = 'Course not found';
            mockCourseRepository.GetCourseByName.mockResolvedValue(undefined);

            await expect(() => courseService.GetCourseByName(courseName)).rejects.toThrow(expectedErrorMessage);
            await expect(() => courseService.GetCourseByName(courseName)).rejects.toThrow(ResourceNotFoundException);

            expect(mockCourseRepository.GetCourseByName).toHaveBeenCalledTimes(2);
            expect(mockCourseRepository.GetCourseByName).toHaveBeenCalledWith(courseName);
        })

        it.each(courseStatistics)('should get course statistics for $courseName from database', async (expectedCourseStatistic) => {
            const courseElement = courses.find(v => v.GetCourseName() === expectedCourseStatistic.GetCourseName()) as Course;
            mockCourseRepository.GetCourseStatistics.mockResolvedValue(expectedCourseStatistic);

            let courseServiceCourses = await courseService.GetCourseStatistics(courseElement);

            expect(courseServiceCourses).toMatchSnapshot();
            expect(mockCourseRepository.GetCourseStatistics).toHaveBeenCalledTimes(1);
        })

    })


    describe('Error path', () => {
        const course = new Course('C4', new Date('2024-04-01'), 5, 40000);
        const student = new Student('Student 1', 20);

        it('should be throw a DatabaseException when add course into database', async () => {
            mockCourseRepository.AddCourse.mockImplementation(() => { throw new DatabaseException('Add course failed') });

            await expect(() => courseService.AddCourse(course)).rejects.toThrow(DatabaseException);
        })

        it('should be throw a UnknownErrorException when add course into database', async () => {
            mockCourseRepository.AddCourse.mockImplementation(() => { throw new UnknownErrorException('Unknown error occured') });

            await expect(() => courseService.AddCourse(course)).rejects.toThrow(UnknownErrorException);
        })

        it('should be throw a PaymentException when add student to course into database', async () => {
            mockPaymentService.isOrderPaid.mockResolvedValue(false);

            await expect(() => courseService.AddStudentToCourse(student, course)).rejects.toThrow(PaymentException);
        })

        it('should be throw a ResourceNotFoundException when the get course by name call with a non-exists course name', async () => {
            const courseName = 'CR11111';
            mockCourseRepository.GetCourseByName.mockImplementation(() => { throw new ResourceNotFoundException('Course not found') });

            await expect(() => courseService.GetCourseByName(courseName)).rejects.toThrow(ResourceNotFoundException);
        })

        it('should be throw a DatabaseException when the get course statistics', async () => {
            mockCourseRepository.GetCourseStatistics.mockImplementation(() => { throw new DatabaseException('Course not found') });

            await expect(() => courseService.GetCourseStatistics(course)).rejects.toThrow(DatabaseException);
        })

    })

})