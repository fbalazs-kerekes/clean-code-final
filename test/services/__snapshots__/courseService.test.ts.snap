// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseService test Happy path should get all course from database 1`] = `
[
  Course {
    "costInHuf": 10000,
    "courseName": "CR1",
    "lengthInWeeks": 5,
    "startDate": 2024-01-01T00:00:00.000Z,
    "students": [],
  },
  Course {
    "costInHuf": 20000,
    "courseName": "CR2",
    "lengthInWeeks": 4,
    "startDate": 2024-02-01T00:00:00.000Z,
    "students": [],
  },
  Course {
    "costInHuf": 30000,
    "courseName": "CR3",
    "lengthInWeeks": 3,
    "startDate": 2024-03-01T00:00:00.000Z,
    "students": [],
  },
]
`;

exports[`CourseService test Happy path should get course statistics for CR1 from database 1`] = `
CourseStatistics {
  "courseName": "CR1",
  "lastAccessed": 2024-01-02T10:10:00.000Z,
  "lecturesCompleted": 50,
  "progress": 30,
  "totalLectures": 100,
}
`;

exports[`CourseService test Happy path should get course statistics for CR2 from database 1`] = `
CourseStatistics {
  "courseName": "CR2",
  "lastAccessed": 2024-02-02T10:10:00.000Z,
  "lecturesCompleted": 50,
  "progress": 30,
  "totalLectures": 200,
}
`;

exports[`CourseService test Happy path should get course statistics for CR3 from database 1`] = `
CourseStatistics {
  "courseName": "CR3",
  "lastAccessed": 2024-03-02T10:10:00.000Z,
  "lecturesCompleted": 50,
  "progress": 30,
  "totalLectures": 150,
}
`;
